fix namespace miss bug

WanLinghao opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This  patch fixes  the namespace miss problems.
I am not sure if this is the correct way it should be fixed.
Just offer a solution.
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #62059

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

CaoShuFeng commented:
/ok-to-test
/assign

WanLinghao commented:
@smarterclayton PTAL

smarterclayton commented:
All objects should implement metadata.Object, even unstructured, so I'm not sure what your fix is trying to do here.

WanLinghao commented:
@juanvallejo @smarterclayton 
Hello, from my opinoin,  this bug was caused from the function
https://github.com/kubernetes/kubernetes/blob/5648200571889140ad246feb82c8f80a5946f167/pkg/kubectl/cmd/resource/get.go#L602

This function should unpack the JSON stored in table.Rows[i].Object.Raw, turning it into the correct object type, and storing it in the table.Rows[i].Object.Object． However it didn't complete this job.

On the other words, the table.Rows[i].Object.Object field in the return value of this function is always empty.This cause the bug.

juanvallejo commented on pkg/printers/humanreadable.go:
```diff
@@ -442,11 +443,15 @@ func DecorateTable(table *metav1beta1.Table, options PrintOptions) error {
-				if acc, err := meta.Accessor(obj); err == nil {
-					m = acc
-				}
+			metaData := &metav1beta1.PartialObjectMetadata{}
```
partialobject metadata is only included in certain table responses from the server. This needs to work with all cases.

A more consistent solution should be for the server to always include the "namespace" column in its columnDefinitions

WanLinghao replied:
@juanvallejo always include the "namespace" for the server is  the solution I figure out in the first place.But I don't think it is best solution.Since currently, function below controls which fields 
the table should include. We can see that the param is fix. So if we modify this , is there any chance that we could influence other invoker? 
https://github.com/kubernetes/kubernetes/blob/a5c3c8d16c74d1fb710a88497f970ef6045f90d4/pkg/printers/storage/storage.go#L30

And for deal with all cases from the server,as far as I can see, code below contains all the possible type the table response may include. 
https://github.com/kubernetes/kubernetes/blob/a5c3c8d16c74d1fb710a88497f970ef6045f90d4/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go#L107

smarterclayton commented:
A client that wants to access namespace should be requesting partial object meta or the full object.  Because namespace is contextual, clients should construct it.

WanLinghao commented:
@smarterclayton  So we should handle this from client side. What about add a switch code at 
https://github.com/kubernetes/kubernetes/blob/5648200571889140ad246feb82c8f80a5946f167/pkg/kubectl/cmd/resource/get.go#L602
To Unmarshal partial object meta or the full object respectively.
What do you think about this solution.

WanLinghao commented:
/test pull-kubernetes-e2e-gce

WanLinghao commented:
@smarterclayton @juanvallejo  I noticed that even if  a full object can be transformed into partial object meta directly by Unmarshal() function.
Is that acceptable?
Since I am not clear with how to deal with full object, so this is as far as I can get

juanvallejo commented:
@WanLinghao by default, a table response from the server only includes partial object metadata within each row. To obtain the full object, the `includeObject=Object` parameter must be specified

WanLinghao committed:
1.fix kubectl get * --all-namespaces
 	  namespace miss error
	2.also add a test case
	modified:   pkg/kubectl/cmd/get/get.go
	modified:   hack/make-rules/test-cmd-util.sh

WanLinghao commented:
/retest

juanvallejo commented:
@WanLinghao  
> Hello, I have updated my patch, I think we can just use unstructured Object here.
Since it can simplify code and compatible with user-defined CRD resource.
WDYT?

Thanks, although your approach does makes sense, it is fixing a temporary workaround that we eventually would rather stop using. The reason the `decodeIntoTable` method is using the json package to unmarshal that runtime.Object into a Table, is because there is currently no codec in our scheme to do this for us. I believe that adding a codec capable of properly "decoding" an object into a Table would be the more long-term approach to this.

liggitt commented:
this fix makes sense to me. unmarshal into `metav1.Table` stops short of decoding RawExtension objects and just leaves them with Raw bytes. Looping and decoding the PartialObjectMetadata objects so that name/namespace can be retrieved from them makes sense to me.

Needs rebase, then LGTM

liggitt commented:
should also add a test to ensure this doesn't regress in the future

WanLinghao commented:
/test pull-kubernetes-integration

liggitt commented on hack/make-rules/test-cmd-util.sh:
```diff
@@ -4474,6 +4474,11 @@ run_kubectl_all_namespace_tests() {
+  ### Check --all-namespaces option shows namespaces
+  # Command
+  output_message=$(kubectl get pods --all-namespaces "${kube_flags[@]}")
+  kube::test::if_has_string "${output_message}" "dddd"
```
where is this dddd namespace coming from?

WanLinghao replied:
@liggitt  sorry, since I am not very clear with how this test works. What I am trying to do here is a test. Try to figure out  how to test if namespace showed correctly.

WanLinghao replied:
@liggitt  thanks, done

WanLinghao commented:
/test pull-kubernetes-verify

WanLinghao commented:
/test pull-kubernetes-verify

WanLinghao commented:
@liggitt Done PTAL

WanLinghao commented:
@liggitt Done PTAL

WanLinghao commented:
@soltysh  if you do not mind，plz unhold This patch

juanvallejo commented:
/retest
